### ISSUE TITLE
`InlineBlocks` pass crashes with `(signal: 11, SIGSEGV: invalid memory reference)`

### ISSUE THREAD

**Issue by greenhat** [0xPolygonMiden/compiler#85]
Discovered in #84 in `test_rust_comp::rust_array()` test.

Un-ignore the test to reproduce.

The IR passes are defined in https://github.com/0xPolygonMiden/compiler/blob/430ef78692174b28de18830e36ffaad1a91462e2/tests/integration/src/compiler_test.rs#L495-L504. The crash goes away if `InlineBlocks` pass is omitted(commented out).

I've also reproduced it with rustc 1.74.1 and the latest nightly. I suspect the `unsafe` code that operates the `intrusive_collections`.

FWIW, here the Wasm:
```wasm
(module
  (type (;0;) (func (param i32 i32) (result i32)))
  (type (;1;) (func (result i32)))
  (func $sum_arr (;0;) (type 0) (param i32 i32) (result i32)
    (local i32)
    i32.const 0
    local.set 2
    block ;; label = @1
      local.get 1
      i32.eqz
      br_if 0 (;@1;)
      loop ;; label = @2
        local.get 0
        i32.load
        local.get 2
        i32.add
        local.set 2
        local.get 0
        i32.const 4
        i32.add
        local.set 0
        local.get 1
        i32.const -1
        i32.add
        local.tee 1
        br_if 0 (;@2;)
      end
    end
    local.get 2
  )
  (func $__main (;1;) (type 1) (result i32)
    i32.const 1048576
    i32.const 5
    call $sum_arr
    i32.const 1048596
    i32.const 5
    call $sum_arr
    i32.add
  )
  (memory (;0;) 17)
  (global $__stack_pointer (;0;) (mut i32) i32.const 1048576)
  (global (;1;) i32 i32.const 1048616)
  (global (;2;) i32 i32.const 1048624)
  (export "memory" (memory 0))
  (export "sum_arr" (func $sum_arr))
  (export "__main" (func $__main))
  (export "__data_end" (global 1))
  (export "__heap_base" (global 2))
  (data $.rodata (;0;) (i32.const 1048576) "\01\00\00\00\02\00\00\00\03\00\00\00\04\00\00\00\05\00\00\00\06\00\00\00\07\00\00\00\08\00\00\00\09\00\00\00\0a\00\00\00")
)
```

**Comment by bitwalker** [comment]
I suspect I know what is causing this, but I'll have to take a look when I have some time to sit down and walk through where exactly things are blowing up to confirm. Might do so this week if I have some spare time.

**Comment by greenhat** [comment]
It's not blocking me, so there's no rush. Here is a HIR before calling `InlineBlocks`:
```rust
module noname

const $0 = 0x00100000;
const $1 = 0x00100028;
const $2 = 0x00100030;

global external @__stack_pointer : i32 = $0 { id = 0 };
global external @gv1 : i32 = $1 { id = 1 };
global external @gv2 : i32 = $2 { id = 2 };

pub fn __main() -> i32 {
block0:
    v1 = const.i32 1048576 : i32;
    v2 = const.i32 5 : i32;
    v3 = call noname::sum_arr(v1, v2) : i32;
    v4 = const.i32 1048596 : i32;
    v5 = const.i32 5 : i32;
    v6 = call noname::sum_arr(v4, v5) : i32;
    v7 = add.wrapping v3, v6 : i32;
    br block1(v7);

block1(v0: i32):
    ret v0;
}

pub fn sum_arr(i32, i32) -> i32 {
block0(v0: i32, v1: i32):
    v3 = const.i32 0 : i32;
    v4 = const.i32 0 : i32;
    v5 = eq v1, 0 : i1;
    v6 = cast v5 : i32;
    v7 = neq v6, 0 : i1;
    condbr v7, block7, block3;

block7:
    br block11;

block11:
    br block12;

block12:
    ret v4;

block3:
    br block4(v0, v4, v1);

block4(v8: i32, v12: i32, v16: i32):
    v9 = cast v8 : u32;
    v10 = inttoptr v9 : *mut i32;
    v11 = load v10 : i32;
    v13 = add.wrapping v11, v12 : i32;
    v14 = const.i32 4 : i32;
    v15 = add.wrapping v8, v14 : i32;
    v17 = const.i32 -1 : i32;
    v18 = add.wrapping v16, v17 : i32;
    v19 = neq v18, 0 : i1;
    condbr v19, block8, block6;

block8:
    br block4(v15, v13, v18);

block5:
    br block9;

block9:
    br block10;

block10:
    ret v13;

block6:
    br block5;
}
```

**Comment by bitwalker** [comment]
This is fixed in #86, so once that is merged and you've rebased on it, you can re-enable that test, update its expect output, and you should be good to go!

**Comment by greenhat** [comment]
Yep, it's working. I enabled the test in https://github.com/0xPolygonMiden/compiler/pull/84/commits/5cedd081cb00afebc4babd665fb0b05d6cf2cd76
